smolyay/surrogate.py

```diff
@@ -117,430 +72,694 @@ def data(self):
-            else:
-                output_shape = input_shape
-            gradient = numpy.reshape(gradient, output_shape)
-        return gradient
 
-    def __call__(self, x):
+class ProductSetSurrogate(BaseSurrogate):
+    r"""Create a surrogate to approximate a complex function.
+
+    Depending on the dimensionality (number of independent variables),
+    sampling method, and combination of basis functions, a surrogate
+    model can be generated that approximates a set of data.
+    ``domain`` is the domain of the function to be approximated.
+    ``basis_sets`` is a list of :class:BasisFunctionSets that describe
+    the basis functions used to construct terms in the surrogate function,
+    with the :class:BasisFunctionSet at index i in the list being the set of
+    basis functions used for the ith dimensional variable.
+    ``alpha`` is the lambda parameter used by Lasso and Ridge Regression
+    ``regression`` is the method of regression used in the event that the
+    number of terms does not match the number of sample points. Ridge
+    regression and lasso regression are available, as well as finding
+    the solution of least squares. If the basis functions return complex
+    values, then solving for least squares will be used instead of ridge
+    or lasso.
+
+    :attr:`points` stores the points that are used for sampling.
+    :attr:`data` stores the output of the true function at the sampled points.
+    :attr:`index_combinations` describes the combination of basis function
+    used to construct the terms of the surrogate, where each row is a term
+    represented by a list of size ``num_dimensions`` that give the index of
+    the basis functions from the :class:BasisFunctionSets in `basis_sets`
+    that make up a given term.
+    :attr:`coefficients` is the list of coefficients for each term in the
+    surrogate equation. These cefficients are determined in :meth:`fit`.
+    :meth:`train`, generates a trained surrogate model given a function and
+    the points to sample at.
+    :meth:`fit` computes the coefficients based on a set of data
+    (function values at transformed grid points).
+    Once the surrogate is constructed, one can evaluate the surrogate
+    through :meth:`predict(x)` and its gradient can be evaluated
+    through :meth:`gradient(x)`.
+    :meth:`_create_terms`, is an abstract method for generating the terms
+    described in :attr:`index_combinations`.
+
+    Parameters
+    ----------
+    domain: numpy.ndarray
+        the domain of the surrgoate
+
+    basis_sets: list of :class:BasisFunctionSet
+        the set of basis functions used to combine terms
+
+    alpha: float, default 1e-10
+        the regression parameter used in Ridge and Lasso regression
+
+    regression : ["ridge", "lasso", "lstsq"], default "ridge"
+        the regression method if the number of points and terms don't match
+    """
+
+    def __init__(self, domain, basis_sets, alpha=1e-10, regression="ridge"):
```

Comment:
I don't think this is the best interface for implementing regularization, since it defaults to using it and introduces a parameter `alpha` that is not always meaningful. I would rather have a parameter `regularization=None` (meaning don't regularize), then add `L1Regularization` and `L2Regularization` helper classes for configuring regularization.

Reply:
The function `numpy.linalg.lstsq` has an extra parameter `rcond` that's used as a cutoff for small singular values of the matrix. We set it to None, but should the user be able to access this variable?

The L1 and L2 regularization are handled by either a `sklearn.linear_model.Ridge` or `sklearn.linear_model.Lasso` object. The parameter `regularization` can be `None` or any child of sklearn's `LinearModel` class. We can be even less strict and have `regularization` either be `None` or anything that has a fit method that returns itself and has a `coef_` attribute. If it's `None`, try `numpy.linalg.solve` or `numpy.linalg.lstsq`. If it's not `None` try the `regularization`'s fit and use `numpy.linalg.lstsq` if that doesn't work.

If we add helper classes, they would be essentially doing the same job that `Ridge` and `Lasso` already do. The only reason to add them is for complex numbers since as far as I can tell, none of sklearn's regressors support complex numbers. I'm not sure if that's something we want to tackle.

Reply:
> The function `numpy.linalg.lstsq` has an extra parameter `rcond` that's used as a cutoff for small singular values of the matrix. We set it to None, but should the user be able to access this variable?

I don't think that is necessary until we have a use case for it.

> The L1 and L2 regularization are handled by either a `sklearn.linear_model.Ridge` or `sklearn.linear_model.Lasso` object.

This is an internal implementation detail, the user shouldn't have to care about how we are doing the regularization. We should just allow the user to configure a minimal set of parameters, and then do the calculation for them.

> If we add helper classes, they would be essentially doing the same job that Ridge and Lasso already do. The only reason to add them is for complex numbers since as far as I can tell, none of sklearn's regressors support complex numbers. I'm not sure if that's something we want to tackle.

The helper classes are for configuration, not doing the calculation. The surrogate can raise an error if regularization is requested with complex numbers, if this is not possible.

Reply:
So the L1 and L2 helper classes will just be containers for sklearn.linear_model.Ridge and sklearn.linear_model.Lasso objects like the random point sets? Should the surrogate require regularization be one of those helper classes or should it accept whatever LinearModel the user wants to put in? If a hypothetical future user wants to use a regularization method from sklearn that isn't Ridge or Lasso, should they have to define an entirely new helper class for that?

>  The surrogate can raise an error if regularization is requested with complex numbers, if this is not possible.

Ridge and Lasso already raise those errors. If regularization is handled by whatever the `regularization` parameter is, then shouldn't it be the thing that raises the error? The surrogate doesn't know if the `regularization` object supports complex numbers, that's its problem.

Reply:
No, they will be objects that contain data to help configure the surrogate. The surrogate is responsible for the regularization as part of its own training.
```py
class L2Regularization:
    def __init__(self, alpha):
        self.alpha = alpha

TensorProductSurrogate(..., regularization=L2Regularization(1e-2))
```
then inside the actual fit method
```py
if isinstance(self.regularization, L2Regularization):
    # do ridge regression
```

Reply:
It works!

Reply:
Looking at the code for the linear models from sklearn, having one of those be a parameter risks that linear model saving a copy of the basis matrix which, if you don't want to use it for any kind of self-assessment, would add a lot of memory to the surrogate object. So I see the logic behind baking in what types of regularizations are allowed in order to prevent a high-memory-cost object from sticking around.